### ISSUE TITLE
Why fix the weight on regression loss?

### ISSUE THREAD

**Issue by shsjxzh** [Zzh-tju/DIoU-SSD-pytorch#21]
In the experiment setting, the paper said that the weight on regression loss is fixed as 5. However, this is unfair to a certain extent, because DIOU and CIOU is larger than IOU, so in the total loss, DIOU and CIOU will naturally occupy a larger portion than IOU loss, which may contribute to the performance gain. In fact, increasing the weight of IOU regression loss can improve the performance a lot.

From my experiment, if I run the original setting for IOU loss for 240 epoch (a little longer than the original setting)，I will achieve 50.21 for AP; If I increase the weight from 5 to 15.3125, it will rise to 52.34, which is higher than original CIOU. 

In conclusion, my question is: can this experiment really support that CIOU and DIOU is better than IOU on SSD?

**Comment by Zzh-tju** [comment]
First, SSD sets the prior boxes which has IoU>0.5 with GT as the positive samples. It results in all the regression cases happened in the basin of Fig 4 in our paper so that all thses IoU-based loss may be similar in some extend. You can refer to GIoU paper [https://arxiv.org/pdf/1902.09630.pdf](url) and see their experiment results on Faster/Mask R-CNN.

Second, the reason why we set regression loss weight as 5 is due to : 
1) we followed [https://github.com/JaryHuang/awesome_SSD_FPN_GIoU](url). 
2) Increasing the loss weight is equivalent to increase the learning rate for regression, it may cause unstable training and make them not so good to compare, since the AP floats too much. I did parameter search for regression loss weight of course, but see sometimes IoU did best, sometimes IoU did worst, sometimes GIoU did worst. D/CIoU‘s’ performance is usually not the worst, medium to high. And this phenomenon is also in line with the idea of simulation experiment that we mentioned in the paper. We can't expect which kind of loss has the highest performance all the time. However, if it performs well in simulation experiment, it will also perform well with a large probability in real training. We got AP 53.34 for CIoU in weight 16 but since we could not get a satisfied result for GIoU, so we gave up that result and it was not good for comparison because of training float. 

Third, if you output the loss for IoU, d²/c² (distance loss) and v (aspect retio loss), you will see the proportion of IoU loss is very large than the latter two. 

Fourth, if we continue to increase regression loss weight, AP will begin to decline after a certain extent, because the proportion of regression loss has been too large and affected the learning of other tasks. The same conclusion can be seen in Libra R-CNN [https://arxiv.org/pdf/1904.02701.pdf](url)

**Comment by shsjxzh** [comment]
Really thanks for your detail reply! It provides much enlightenment.
1) I will compare the total sum of IoU, d²/c² (distance loss) and v (aspect retio loss). This sounds a reasonable explanation. 
2) These detail experiment information helps a lot. I understand your concern, but sharing the result of varied weight will make your support stronger. I think GIOU has its own problem and it is normal that sometimes it will hurt the performance. 

The following lines may be too ideal: we, the researcher, all want to share beautiful results on paper, with some methods like cherry picking. However, if you can share more original results, even in the appendix, will help others a lot and contribute to the betterment of community. Again, really thanks for your detail reply!

**Comment by Zzh-tju** [comment]
Wish these will help you. And more improvement for this repo is coming soon, within the next one week

**Comment by shsjxzh** [comment]
Looking forward to it ~
